## Modelo matematico del Titanic 
comenzamos con las importacion y el convertir nuestro tarin.csv a un DataFrame para mosrar una parte de su contenido

In [1]:
import numpy as np
import pandas as pd
df = pd.read_csv('train.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [2]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [3]:
# Eliminacion de variables
del df['PassengerId']
del df['Name']

In [4]:
# Mostrar si hay valores faltantes y cuantos son
df.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

De 891 muestras, 687 tienen valores nulos para la variable Cabin. Hay demasiados valores faltantes para que esta variable pueda ser utilizada. La borraremos.

In [5]:
# Eliminacion de otra variable
del df['Cabin']

En el caso de la variable Embarked, como es una variable categórica, miraremos los recuentos de cada categoría y sustituiremos los 2 valores nulos por la categoría que tiene más elementos.

In [6]:
df['Embarked'].value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

El valor más frecuente para Embarked es 'S', así que lo usaremos para reemplazar los valores nulos.

In [7]:
df['Embarked'].loc[pd.isnull(df['Embarked'])] = 'S'

/tmp/ipykernel_13181/1928886212.py:1: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df['Embarked'].loc[pd.isnull(df['Embarked'])] = 'S'
/tmp/ipykernel_13181/1928886212.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a

In [8]:
# En cuanto a la edad, sustituiremos los valores faltantes con la edad promedio
mean_age_train = np.mean(df['Age'].loc[pd.isnull(df['Age']) == False].values)
df['Age'].loc[pd.isnull(df['Age'])] = mean_age_train

/tmp/ipykernel_13181/1024468475.py:3: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df['Age'].loc[pd.isnull(df['Age'])] = mean_age_train
/tmp/ipykernel_13181/1024468475.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of 

Con lo anterior nos decisimos de los valores nulos
Ahora calcularemos y almacenaremos la tarifa media en caso de que la necesitemos en los datos de las pruebas.

In [9]:
mean_fare_train = np.mean(df['Fare'].loc[pd.isnull(df['Fare']) == False].values)

# Codificación ordinal
La codificación ordinal es una forma de convertir una variable categórica a números asignando a cada categoría un número. 

In [10]:
# Copia de seguridad de nuestro actual formato de datos para su uso futuro.
df_bkp = df.copy()
from sklearn.preprocessing import OrdinalEncoder
df['Sex'] = OrdinalEncoder().fit_transform(df['Sex'].values.reshape((-1, 1)))
df['Ticket'] = OrdinalEncoder().fit_transform(df['Ticket'].values.reshape((-1, 1)))
df['Embarked'] = OrdinalEncoder().fit_transform(df['Embarked'].values.reshape((-1, 1)))

# Visualizaciones
Ahora visualizaremos nuestros datos para ver cómo varía la distribución de nuestras variables entre las clases sobrevivientes y no sobrevivientes.

In [11]:
import matplotlib.pyplot as plt
from IPython.display import display, Markdown

def show(txt):
    # this function is for printing markdown in jupyter notebook
    display(Markdown(txt))

for i in range(1, 9):
    show(f'### {df.columns[i]}')
    f, (survived, not_survived) = plt.subplots(1, 2, sharey=True, figsize=(18, 8))
    survived.hist(df.iloc[np.where(df['Survived'] == 1)[0], i])
    survived.set_title('Survived')

    not_survived.hist(df.iloc[np.where(df['Survived'] == 0)[0], i])
    not_survived.set_title('Not Survived')
    plt.show()

ModuleNotFoundError: No module named 'matplotlib'

# Corriendo aprendizaje automático en este formato de datos
Ahora, probaremos un par de métodos de aprendizaje automático en nuestro dataset para ver qué resultados obtenemos. Los métodos de aprendizaje automático que usaremos son:

    -Regresión logística
    -Support vector machines
    -Árbol de decisión
    -K Vecinos más cercanos
    -Perceptrón multicapa


In [ ]:
# En lugar de elegir un conjunto de validación fijo para estimar la precisión del conjunto de pruebas, utilizaremos una validación cruzada (5 veces) con el cross_val_score de Scikit-Learn, que devuelve una matriz con las puntuaciones de cada iteración de validación cruzada.
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier


X = df.iloc[:, 1:].values
y = df.iloc[:, 0].values


# Logistic Regression
lr = LogisticRegression()
lr_score = np.mean(cross_val_score(lr, X, y))
print(f'Logistic Regression: {lr_score}')


# Support Vector Machine
svc = SVC()
svc_score = np.mean(cross_val_score(svc, X, y))
print(f'Support Vector Machine: {svc_score}')


# Decision Tree
dtc = DecisionTreeClassifier()
dtc_score = np.mean(cross_val_score(dtc, X, y))
print(f'Decision Tree: {dtc_score}')


# K Nearest Neighbors
knc = KNeighborsClassifier()
knc_score = np.mean(cross_val_score(knc, X, y))
print(f'K Nearest Neighbors: {knc_score}')


# Multi-Layer Perceptron
mlpc = MLPClassifier()
mlpc_score = np.mean(cross_val_score(mlpc, X, y))
print(f'Multi-Layer Perceptron: {mlpc_score}')

/home/hernan_so5/venv/lib/python3.13/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/hernan_so5/venv/lib/python3.13/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.htm

Logistic Regression: 0.793496955621116
Support Vector Machine: 0.6768062268533048
Decision Tree: 0.7733098989391751
K Nearest Neighbors: 0.6869374176134581
Multi-Layer Perceptron: 0.7418931642709184


# Ingeniería de características
restauraremos la copia de seguridad que hicimos antes de aplicar la codificación ordinal.

In [ ]:
df = df_bkp.copy()

# Codificación única (One Hot Encoding)
ahora queremos aplicar one-hot encoding a nuestras variables categóricas. Con one-hot encoding, en lugar de asignar a cada clase un número, asignamos un vector lleno de 0’s excepto para una posición específica de esa clase en la que ponemos un 1. Es decir, transformamos cada clase en una variable por sí misma que toma el valor 0 o 1. Aplicaremos one-hot encoding en las variables Pclass, Sex, Ticket y Embarked usando el OneHotEncoder de Scikit-Learn.

In [ ]:
from sklearn.preprocessing import OneHotEncoder
import numpy as np

# Pclass
pclass_transf = OneHotEncoder(sparse_output=False, dtype=np.uint8, handle_unknown='ignore')
pclass_transf.fit(df['Pclass'].values.reshape((-1, 1)))
pclass = pclass_transf.transform(df['Pclass'].values.reshape((-1, 1)))
df['Pclass0'] = pclass[:, 0]
df['Pclass1'] = pclass[:, 1]
df['Pclass2'] = pclass[:, 2]
del df['Pclass']

# Sex
gender_transf = OneHotEncoder(sparse_output=False, dtype=np.uint8, handle_unknown='ignore')
gender_transf.fit(df['Sex'].values.reshape((-1, 1)))
gender = gender_transf.transform(df['Sex'].values.reshape((-1, 1)))
df['Male'] = gender[:, 0]
df['Female'] = gender[:, 1]
del df['Sex']

# Ticket 
ticket_transf = OneHotEncoder(sparse_output=False, dtype=np.uint8, handle_unknown='ignore')
ticket_transf.fit(df['Ticket'].values.reshape((-1, 1)))
ticket = ticket_transf.transform(df['Ticket'].values.reshape((-1, 1)))
for i in range(ticket.shape[1]):
    df[f'Ticket{i}'] = ticket[:, i]
del df['Ticket']

# Embarked
embarked_transf = OneHotEncoder(sparse_output=False, dtype=np.uint8, handle_unknown='ignore')
embarked_transf.fit(df['Embarked'].values.reshape((-1, 1)))
embarked = embarked_transf.transform(df['Embarked'].values.reshape((-1, 1)))
for i in range(embarked.shape[1]):
    df[f'Embarked{i}'] = embarked[:, i]
del df['Embarked']

/tmp/ipykernel_24540/2476900192.py:26: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[f'Ticket{i}'] = ticket[:, i]
/tmp/ipykernel_24540/2476900192.py:26: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[f'Ticket{i}'] = ticket[:, i]
/tmp/ipykernel_24540/2476900192.py:26: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe 

# Escalando a rangos entre [0, 1]

También queremos escalar las variables numéricas a un rango [0, 1]. Para ello utilizaremos MinMaxScaler que escala las variables de manera que el valor mínimo se mueva a 0, el máximo a 1 y los valores intermedios se escalen consecuentemente entre 0 y 1. 

In [ ]:
# Aplicaremos esta transformación a las variables Age, SibSp, Parch, Fare.

from sklearn.preprocessing import MinMaxScaler

age_transf = MinMaxScaler().fit(df['Age'].values.reshape(-1, 1))
df['Age'] = age_transf.transform(df['Age'].values.reshape(-1, 1))

sibsp_transf = MinMaxScaler().fit(df['SibSp'].values.reshape(-1, 1))
df['SibSp'] = sibsp_transf.transform(df['SibSp'].values.reshape(-1, 1))

parch_transf = MinMaxScaler().fit(df['Parch'].values.reshape(-1, 1))
df['Parch'] = parch_transf.transform(df['Parch'].values.reshape(-1, 1))

fare_transf = MinMaxScaler().fit(df['Fare'].values.reshape(-1, 1))
df['Fare'] = fare_transf.transform(df['Fare'].values.reshape(-1, 1))

# Haciendo aprendizaje automático en este nuevo formato de datos

Ahora ejecutaremos los mismos algoritmos de aprendizaje de la máquina en este nuevo formato de datos para ver qué resultados obtenemos.

In [ ]:
X = df.iloc[:, 1:].values
y = df.iloc[:, 0].values

# Logistic Regression
lr = LogisticRegression()
lr_score = np.mean(cross_val_score(lr, X, y))
print(f'Logistic Regression: {lr_score}')

# Support Vector Machine
svc = SVC()
svc_score = np.mean(cross_val_score(svc, X, y))
print(f'Support Vector Machine: {svc_score}')

# Decision Tree
dtc = DecisionTreeClassifier()
dtc_score = np.mean(cross_val_score(dtc, X, y))
print(f'Decision Tree: {dtc_score}')

# K Nearest Neighbors
knc = KNeighborsClassifier()
knc_score = np.mean(cross_val_score(knc, X, y))
print(f'K Nearest Neighbors: {knc_score}')

# Multi-Layer Perceptron
mlpc = MLPClassifier()
mlpc_score = np.mean(cross_val_score(mlpc, X, y))
print(f'Multi-Layer Perceptron: {mlpc_score}')

Logistic Regression: 0.8136902893729208
Support Vector Machine: 0.8125792480070302
Decision Tree: 0.8339212855439081
K Nearest Neighbors: 0.7980038917833155


/home/hernan_so5/venv/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:781: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/hernan_so5/venv/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:781: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/hernan_so5/venv/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:781: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/hernan_so5/venv/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:781: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Multi-Layer Perceptron: 0.8159312033142928


/home/hernan_so5/venv/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:781: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


# Ajuste de Hiper-Parámetros

In [ ]:
from sklearn.model_selection import GridSearchCV


dtc = DecisionTreeClassifier()
params = {
    'max_depth': list(range(2, 151)),
    'min_samples_split': list(range(2, 15))
}

clf = GridSearchCV(dtc, params)
clf.fit(X, y)
print(f'Best params: {clf.best_params_}')
print(f'Best score: {clf.best_score_}')

Best params: {'max_depth': 142, 'min_samples_split': 13}
Best score: 0.8440273680246062
